src/ast_editor.ml

```diff
@@ -0,0 +1,585 @@
+
+  let get_kind ~document =
+    let relative = relative_document_path ~document in
+    match Caml.Filename.extension relative with
+    | ".ml" -> Structure
+    | ".mli" -> Signature
+    | _ -> Unknown
+
+  let get_pp_path ~(document : TextDocument.t) =
+    let relative = relative_document_path ~document in
+    match project_root_path () with
+    | None ->
+      let (_ : _ Promise.t) =
+        Window.showErrorMessage ~message:"Project root wasn't found." ()
+      in
+      None
+    | Some root ->
+      let build_root = "_build/default" in
+      let fname =
+        match get_kind ~document with
+        | Unknown ->
+          let uri = Uri.toString (TextDocument.uri document) () in
+          raise (User_error (sprintf "File %s has unknown file extension" uri))
+        | Structure -> String.chop_suffix_exn ~suffix:".ml" relative ^ ".pp.ml"
+        | Signature ->
+          String.chop_suffix_exn ~suffix:".mli" relative ^ ".pp.mli"
+      in
+      let ( / ) = Caml.Filename.concat in
+      Some (root / build_root / fname)
+end
+
+let fetch_pp_code ~document =
+  match Pp_path.get_pp_path ~document with
+  | None ->
+    show_message `Error "Unable to find preprocessed file for %s"
+      (Uri.toString (TextDocument.uri document) ());
+    Ok ""
+  | Some path -> (
+    match Ppx_tools.get_reparsed_code_from_pp_file ~path with
+    | Ok code -> Ok code
+    | Error (Io_error err_msg) -> Error err_msg
+    | Error (Read_error err_msg) ->
+      show_message `Error "%s" err_msg;
+      Ok "")
+
+let transform_to_ast instance ~(document : TextDocument.t)
+    ~(webview : WebView.t) =
+  let ast_editor_state = Extension_instance.ast_editor_state instance in
+  let module Dumpast = Ppx_tools.Dumpast in
+  let make_value result =
+    match result with
+    | Ok ast -> ast
+    | Error error_msg -> raise (User_error error_msg)
+  in
+  if Ast_editor_state.get_original_mode ast_editor_state then
+    let origin_json =
+      let text = TextDocument.getText document () in
+      match Pp_path.get_kind ~document with
+      | Structure -> make_value (Dumpast.transform text `Impl)
+      | Signature -> make_value (Dumpast.transform text `Intf)
+      | Unknown -> raise (User_error "Unknown file extension")
+    in
+    send_msg "ast" (Jsonoo.t_to_js origin_json) ~webview
+  else
+    let pp_value =
+      match Pp_path.get_pp_path ~document with
+      | None -> raise (User_error "Project root path wasn't found")
+      | Some path -> (
+        match Ppx_tools.get_preprocessed_ast path with
+        | Error (Io_error err_msg)
+        | Error (Read_error err_msg) ->
+          raise (User_error err_msg)
+        | Ok res ->
+          let pp_json =
+            let pp_code =
+              match fetch_pp_code ~document with
+              | Ok s -> s
+              | Error m -> raise (User_error m)
+            in
+            let lex = Lexing.from_string pp_code in
+            Result.ok_or_failwith
```

Comment:
Why is this `failwith`? Isn't this an error that should be displayed in the WebView? (parsing error of sorts).

Reply:
Changed it to managing `Error` (which should never occur since all of the `Traverse_ast2.lift2`methods returning it are accounted for by `with_reparse` (except if I forgot something).